check for chroot in systemd module

jakeo opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

Checks for chroot and warn when managing service state w/ systemd module.

In a chrooted environment, `result['status']` doesn't exist, so basic service/systemd tasks w/ `state: started` fail.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
systemd

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.2
  config file = /vagrant/ansible.cfg
  configured module search path = [u'/vagrant/library']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

In a chrooted environment, `result['status']` doesn't exist, so basic service/systemd tasks w/ `state: started` fail.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
- name: start
  service:
    name: apparmor
    state: started

```
```
# chroot /build
# systemctl status apparmor
Running in chroot, ignoring request.

```
versus
```
# systemctl status apparmor
● apparmor.service - AppArmor initialization
   Loaded: loaded (/lib/systemd/system/apparmor.service; enabled; vendor preset: enabled)
   Active: active (exited) since Thu 2018-08-09 22:42:17 UTC; 4min 2s ago
     Docs: man:apparmor(7)
           http://wiki.apparmor.net/
  Process: 331 ExecStart=/etc/init.d/apparmor start (code=exited, status=0/SUCCESS)
 Main PID: 331 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/apparmor.service

Aug 09 22:42:16 build-debian systemd[1]: Starting AppArmor initialization...
Aug 09 22:42:17 build-debian apparmor[331]: Starting AppArmor profiles:
Aug 09 22:42:17 build-debian apparmor[331]: Warning failed to create cache: usr.sbin.sssd
Aug 09 22:42:17 build-debian apparmor[331]: .
Aug 09 22:42:17 build-debian systemd[1]: Started AppArmor initialization.
```

jakeo committed:
check for result['status'] in systemd module

samdoran commented:
Rather than silently skipping the state setting and reporting `ok`, I think it would be better warn that the sate wasn't changed because of the `chroot` environment.

cc @bcoca

bcoca commented:
I think we either need to set a status to a value that makes sense or find an alternate way to determine it in a chroot

jakeo commented:
@bcoca I totally agree - this was a super quick fix, I'll put in a specific check for a chroot check, but it's gonna coincide w/ other chroot checks in the codebase. Cool?

bcoca commented:
that is fine, though we should look at putting them in common location in module_utils

jakeo committed:
instead of checking for result['state'], actually check for chroot and warn

samdoran reviewed (changes requested):
After some further discussion, this fix is not quite correct. Changing the failure to a warning with success is misleading. This should be a warning if running in a chroot environment and fail if the module cannot make a change as a result of that.

It could be the module is running in a chroot environment and it _is_ able to make changes (probably rare). In this case, a warning with success is helpful since it lets the user know about the chroot environment, which is helpful information.

bcoca commented:
I would just warn in either case 'target is a chroot, which can be misleading and either give false positives or prevent the advertised init system tools from working'. <= covers both 'false success' and 'failure cause of chroot' ... also can be ignore don 'non false success'.

jakeo committed:
allow systemctl status to work if in a chroot, update warn text

jakeo commented:
@bcoca @samdoran thanks for the feedback and sorry this has taken so long for me to address.

I _think_ I've addressed your comments, please let me know otherwise!

bcoca commented on lib/ansible/modules/system/systemd.py:
```diff
@@ -494,8 +495,12 @@ def main():
-                module.fail_json(msg="Service is in unknown state", status=result['status'])
+                # check for chroot
+                if is_chroot():
+                    module.warn("Target is a chroot. This can lead to false positives or prevent the init system tools from working." % unit)
```
i'm fine with adding the warning, but why change the failure condition?

jakeo replied:
derp - i misunderstood your request for changes

jakeo committed:
simply change warning message

samdoran committed:
Add changelog

samdoran commented:
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backport-pull-request-process) for this to be included in previous versions.
